core/trino-main/src/test/java/io/trino/operator/TestHashAggregationOperator.java

```diff
@@ -698,6 +712,123 @@ public void testMemoryTracking()
+                .addBlocksPage(createRLEBlock(1, 10), createRLEBlock(1, 10))
+                .addBlocksPage(createRLEBlock(2, 10), createRLEBlock(2, 10))
+                .build();
+
+        assertOutputEquals(operatorFactory, operator2Input, operator2Expected);
+    }
+
+    @Test
+    public void testAdaptivePartialAggregationTriggeredOnlyOnFlush()
+            throws Exception
+    {
+        List<Integer> hashChannels = Ints.asList(0);
+
+        HashAggregationOperatorFactory operatorFactory = new HashAggregationOperatorFactory(
+                0,
+                new PlanNodeId("test"),
+                ImmutableList.of(BIGINT),
+                hashChannels,
+                ImmutableList.of(),
+                PARTIAL,
+                ImmutableList.of(LONG_MIN.createAggregatorFactory(PARTIAL, ImmutableList.of(0), OptionalInt.empty())),
+                Optional.empty(),
+                Optional.empty(),
+                10,
+                Optional.of(DataSize.of(16, MEGABYTE)), // this setting makes operator to flush only after all pages
+                joinCompiler,
+                blockTypeOperators,
+                // use 5 rows threshold to trigger adaptive partial aggregation after each page flush
+                Optional.of(new PartialAggregationController(5, 0.8)));
+
+        List<Page> operator1Input = rowPagesBuilder(false, hashChannels, BIGINT)
+                .addSequencePage(10, 0) // first page are unique values, so it would trigger adaptation, but it won't because flush is not called
+                .addBlocksPage(createRLEBlock(1, 2)) // second page will be hashed to existing value 1
+                .build();
+        // the total unique ows ratio for the first operator will be 10/12 so > 0.8 (adaptive partial aggregation uniqueRowsRatioThreshold)
+        List<Page> operator1Expected = rowPagesBuilder(BIGINT, BIGINT)
+                .addSequencePage(10, 0, 0) // we are expecting second page to be squashed with the first
+                .build();
+        assertOutputEquals(operatorFactory, operator1Input, operator1Expected);
+
+        // Second operator using the same factory, reuses PartialAggregationControl, so it will only produce raw pages (partial aggregation is disabled at this point)
+        List<Page> operator2Input = rowPagesBuilder(false, hashChannels, BIGINT)
+                .addBlocksPage(createRLEBlock(1, 10))
+                .addBlocksPage(createRLEBlock(2, 10))
+                .build();
+        List<Page> operator2Expected = rowPagesBuilder(BIGINT, BIGINT)
+                .addBlocksPage(createRLEBlock(1, 10), createRLEBlock(1, 10))
+                .addBlocksPage(createRLEBlock(2, 10), createRLEBlock(2, 10))
+                .build();
+
+        assertOutputEquals(operatorFactory, operator2Input, operator2Expected);
+    }
+
+    private void assertOutputEquals(OperatorFactory operatorFactory, List<Page> input, List<Page> expectedPages)
```

Comment:
you can use one of `io.trino.operator.OperatorAssertion#assertOperatorEquals` instead of creating a new method (see `TestHashJoinOperator`)

Reply:
`OperatorAssertion#assertOperatorEquals` closes the factory and in this case I need the factory to be reused but I extended assertOperatorEquals with  `boolean closeOperatorFactory` so it works now.